sklearn/preprocessing/_data.py

```diff
@@ -765,28 +783,103 @@ def partial_fit(self, X, y=None):
                 raise ValueError(
                     "Cannot center sparse matrices: pass `with_mean=False` "
                     "instead. See docstring for motivation and alternatives.")
-
             sparse_constructor = (sparse.csr_matrix
                                   if X.format == 'csr' else sparse.csc_matrix)
-            counts_nan = sparse_constructor(
+
+            if sample_weight is not None:
+                counts_nan = sparse_constructor(
+                    (np.isnan(X.data), X.indices, X.indptr),
+                    shape=X.shape,
+                    dtype=sample_weight.dtype)
+                if counts_nan.format == 'csr':
+                    counts_nan.data *= sample_weight.repeat(
+                        np.diff(counts_nan.indptr))
+                elif counts_nan.format == 'csc':
+                    counts_nan.data *= sample_weight[counts_nan.row]
+                counts_nan = counts_nan.sum(axis=0).A.ravel()
+            else:
+                counts_nan = sparse_constructor(
                         (np.isnan(X.data), X.indices, X.indptr),
                         shape=X.shape).sum(axis=0).A.ravel()
 
             if not hasattr(self, 'n_samples_seen_'):
-                self.n_samples_seen_ = (
+                if sample_weight is not None:
+
+                    self.n_samples_seen_ = (
+                       sample_weight.sum() - counts_nan).astype(np.int64,
+                                                                copy=False)
```

Comment:
We can no longer use integers here because the weights can be non-integer floating point values. Maybe instead we should probably introduce a new attribute (e.g.  `sample_weight_sums_` that is always of shape `(n_features,)` with a `np.float64` dtype) for simplicity. We can leave the `n_samples_seen_` unchanged (with int64 dtype) when `sample_weight=None` for backward compat but not use it anymore internally.